url7d636e8121_laforge49_JActor2_tgz-pJ8-org_agilewiki_jactor2_core_revisited_EHJ8/src/org/agilewiki/jactor2/core/impl/mtReactors/ReactorMtImpl.java

```diff
@@ -0,0 +1,784 @@
+    @Override
+    public final boolean isRunning() {
+        return running;
+    }
+
+    /**
+     * Returns the logger.
+     *
+     * @return A logger.
+     */
+    public Logger getLogger() {
+        return logger;
+    }
+
+    /**
+     * Returns true when the first phase of closing has begun.
+     *
+     * @return True when the first phase of closing has begun.
+     */
+    protected final boolean startedClosing() {
+        return startClosing;
+    }
+
+    public final boolean isClosing() {
+        return shuttingDown;
+    }
+
+    /**
+     * Removes this ReactorImpl from any closers, closes any requests that are in process,
+     * closes all closeables, closes the outbox, closes the inbox, interrupts the processing
+     * of the current thread and, if the interrupt is not effective, begins hung thread recovery.
+     */
+    @Override
+    public void close() throws Exception {
+        fail(null);
+    }
+
+    public String getReasonForFailure() {
+        return reason;
+    }
+
+    /**
+     * Close the reactor;
+     *
+     * @param _reason The reason why the reactor is being closed,
+     *                or null if not a failure.
+     */
+    @Override
+    public void fail(final String _reason) throws Exception {
+        reason = _reason;
+        closeableImpl.close();
+
+        if (startClosing) {
+            return;
+        }
+        startClosing = true;
+
+        if (closeables != null) {
+            final HashSet<Closeable> hs = new HashSet<Closeable>(closeables);
+            Iterator<Closeable> cit = hs.iterator();
+            while (cit.hasNext()) {
+                final Closeable closeable = cit.next();
+                try {
+                    closeable.close();
+                } catch (final Throwable t) {
+                    if ((closeable != null) && PlantMtImpl.DEBUG) {
+                        getLogger().warn(
+                                "Error closing a "
+                                        + closeable.getClass().getName(), t);
+                    }
+                }
+            }
+            cit = closeables.iterator();
+            while (cit.hasNext()) {
+                final Closeable closeable = cit.next();
+                warn("still has closable: " + this + "\n" + closeable);
+            }
+        }
+
+        shuttingDown = true;
+
+        final PlantMtImpl plantImpl = PlantMtImpl.getSingleton();
+        if ((plantImpl != null) && isRunning()
+                && ((currentRequest == null) || !currentRequest.isComplete())) {
+            timeoutSemaphore = plantImpl.schedulableSemaphore(recovery
+                    .getThreadInterruptMillis(this));
+            final Thread thread = getThreadReference().get();
+            if (thread != null) {
+                thread.interrupt();
+                boolean timeout = false;
+                try {
+                    timeout = timeoutSemaphore.acquire();
+                } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)